### ISSUE TITLE
Update plane feature

### ISSUE THREAD

**Issue by aashish24** [OpenGeoscience/geojs#228]
I want to call it rectFeature and provide the ability to have more than one...

**Comment by manthey** [comment]
It seems like we should move in another direction and allow quadrilateral features that don't have to be rectangles, but would be easier to instantiate as rectangles.  This would make doing interesting map projections easier.  For instance, if you used a sinusoidal projection, square map tiles could be distorted into arbitrary quadrilaterals, and, at high enough zoom levels, the results would look okay.

**Comment by aashish24** [comment]
> It seems like we should move in another direction and allow quadrilateral features that don't have to be rectangles, but would be easier to instantiate as rectangles. This would make doing interesting map projections easier. For instance, if you used a sinusoidal projection, square map tiles could be distorted into arbitrary quadrilaterals, and, at high enough zoom levels, the results would look okay.

+1. Also, we should have this feature create multiple instance of plane or rectangle using single geometry vs individual. Current planeFeature behaves differently than point and line.

**Comment by aashish24** [comment]
@manthey based on our discussions I think we need to do the following: 

- [ ] Rename planeFeature to quadFeature
- [ ] Using geometry instancing to save memory and for performance
- [ ] Fix API so that we can pass coordinates for multiple quads

Let me know if you want to take of this.

**Comment by manthey** [comment]
Currently, in vgl, there are two create*Plane functions, one for untextured (solid color) and one for textured planes.  Neither of these currently support multiple planes.  In geojs, when we create multiple points, we don't use the `vgl.utils.createPoints` function.  Rather, the bulk of the points feature is in geojs, including the shaders.

I can either add this as a `vgl.utils.createQuads` function, or with the bulk of the code in geojs `geo.gl.quadFeature`.  Either way, there will be a `quadFeature` in geojs and new shaders will be needed.  As geojs gets much more thorough testing, my inclination is to put the bulk (possibly all) of the code in geojs.

Opinions?

**Comment by jbeezley** [comment]
I think that we need to define the boundary between vgl and geojs.  I believe the original intent was to use vgl as an abstraction layer around GL, but increasingly the GL code has bled into geojs.  Without clear API with documentation and testing in vgl, it is difficult know where that line should exist.

We should either commit to using a true wrapper around GL or get rid of the idea completely.  If we do want to continue with an abstraction, my preference would be to migrate to an established package like [stackgl](http://stack.gl/packages/) to do it.

**Comment by aashish24** [comment]
> I think that we need to define the boundary between vgl and geojs. I believe the original intent was to use vgl as an abstraction layer around GL

The boundary is the VGL is object oriented GL and nothing more. All of the create methods are there to make things simpler. @manthey I am fine with createQuads in GeoJS since that is not object oriented GL. 

@jbeezley I am not in strong favor of stackgl, I don't see a strong need as of now.

**Comment by jbeezley** [comment]
I see a strong need for documented and tested code.  That's what stackgl provides.

**Comment by aashish24** [comment]
> I see a strong need for documented and tested code. That's what stackgl provides.

@jbeezley that's what we are doing in GeoJS and GeoJS should test its GL features. I don't want us to open another front since we have other priorities at the moment. We can consider using other renderers (cesium, stackgl) in the future since GeoJS api allows us to add new renderers if needed.

**Comment by jbeezley** [comment]
In my opinion, the statement "GeoJS should test its GL features" is unacceptable.  If we don't have the resources to maintain vgl properly, we need to transition away from it.

**Comment by aashish24** [comment]
> In my opinion, the statement "GeoJS should test its GL features" is unacceptable

@jbeezley are you suggesting that the GL features in GeoJS should not be tested by GeoJS? Let's talk more offline.

**Comment by jbeezley** [comment]
I'm saying vgl should be testing in vgl.

**Comment by aashish24** [comment]
> I'm saying vgl should be testing in vgl.

Thanks, what I was suggesting earlier is different, GeoJS needs to test its GL features is what I was referring to. Lets talk more as I think this issue is not the best place to discuss this.

**Comment by manthey** [comment]
On another note, I propose the following interface for `quadFeature`:

Class construction args:
```
{
  style: {
    color: {r: 1.0, g: 1.0, b: 1.0},
    opacity: 1,
    depth: 0,
    drawOnAsyncResourceLoaded: true,
    image: function (d) {
      return d.image;
    },
    position: function (d) {
      return d;
    }
  }
}
```
If image is undefined or null, then the quad is a solid color (with opacity).  Otherwise, it is a textured quad with the image mapped corner-to-corner to the quad (thus, if the quad is not rectangular, skewing the texture appropriately -- this will allow us to better support arbitrary transforms).

Each quad's geometry can be defined by a variety of optional elements in the position object, like so:
```
{
  ul: {x: <x>, y: <y>[, z: <z>]},
  ll: {x: <x>, y: <y>[, z: <z>]},
  ur: {x: <x>, y: <y>[, z: <z>]},
  lr: {x: <x>, y: <y>[, z: <z>]},
  origin: {x: <x>, y: <y>[, z: <z>]}
}
```
where at least two opposite corners are needed.  If a corner is excluded, then it is assumed to be rectilinear with the appropriate coordinates of the adjacent corners.  If `origin` is excluded, it is the same as corner ll.  If no z value is specified, the depth from the style is used.  Instead of a point being specified with x, y[, z], it can be an array of two or three values.

Since the corner names are arbitrary, we could, instead, specify `corners: (array of points)`, but this makes it harder to let the interface specify partial data as a convenience for rectangles.

We really need to document our existing features better -- the API docs should be sufficient to learn about things like `drawOnAsyncResourceLoaded`, and they aren't.

Questions:

- Does anyone want a different interface for the feature?

- Is there any reason to have a fallback color for a quad while an image is loading?  If so, should the opacity of that fill color be independent of the opacity of the texture?

**Comment by manthey** [comment]
If an image always maps to the four corners, the origin becomes pretty meaningless -- we really just need a way to say which is the first corner and what orientation should the image have.  For a `corners: (array of either two or four points)` the origin could always be the first point and the direction would be assumed in the two point case and explicit in the four point case.

**Comment by jbeezley** [comment]
I would prefer an API that takes an object as you proposed rather than an array.

Having a fallback color (or better yet a fallback texture) is a very nice feature to have.  Google maps uses a kind of grid-like texture that lets you see things like panning and zooming even when no tiles are loaded.  We could improve the visual quality of our examples a lot by coordinating the style of the tileset with the temporary background.  Most online examples I see use a background that looks similar to land in the tiles.

I can't think of any real compelling reasons to disconnect the fall-back opacity with the main tile opacity.

**Comment by aashish24** [comment]
> style: {
>     color: {r: 1.0, g: 1.0, b: 1.0},
>     opacity: 1,
>     depth: 0,
>     drawOnAsyncResourceLoaded: true,
>     image: function (d) {
>       return d.image;
>     },
>     position: function (d) {
>       return d;
>     }
>   }

do we want to allow setting individual colors or opacity for each? I think that could be useful in some cases but not in majority of them.

**Comment by aashish24** [comment]
> style: {
> color: {r: 1.0, g: 1.0, b: 1.0},
> opacity: 1,
> depth: 0,
> drawOnAsyncResourceLoaded: true,
> image: function (d) {
> return d.image;
> },
> position: function (d) {
> return d;
> }
> }

Assuming that you can provide a function like other features then my question could be scratched.

**Comment by manthey** [comment]
That can be done through our style interface when a function is specified instead of a fixed value or object.

**Comment by aashish24** [comment]
> {
>   ul: {x: <x>, y: <y>[, z: <z>]},
>   ll: {x: <x>, y: <y>[, z: <z>]},
>   ur: {x: <x>, y: <y>[, z: <z>]},
>   lr: {x: <x>, y: <y>[, z: <z>]},
>   origin: {x: <x>, y: <y>[, z: <z>]}
> }

+1 



> Is there any reason to have a fallback color for a quad while an image is loading? 
It might be good to have this option. For example, even if a image fails to load the user may still want to show that tile as missing data (I have seen such use-cases in past). 

If so, should the opacity of that fill color be independent of the opacity of the texture?
I think we can use the same opacity. There will be very few use-cases where we want them to be different. A option for modularity would be nice where you can combine the quad color with texture color.

**Comment by manthey** [comment]
We could add `previewColor` and `previewImage` for quads with an image that hasn't loaded yet.  For our tile layers, this probably isn't desirable (at least when using lower tiles), as we want missing tiles to be clear, as the lower levels are better than a fallback value.

**Comment by manthey** [comment]
Blending color with texture adds needing another property on how the blending is done.  It adds a fair amount to the complexity of things (especially testing).  I'd rather not add that unless we have a use case for it.

**Comment by aashish24** [comment]
I think calling color simply woudl be fine instead of previewColor. 

> Blending color with texture adds needing another property on how the blending is done. It adds a fair amount to the complexity of things (especially testing). I'd rather not add that unless we have a use case for it.

Sure, that's fine, we can do some other time, not required right now unless we need it for bio-imaging.

**Comment by manthey** [comment]
If we show `color` while waiting for tiles to load, then for our standard tile use case, we'll need to explicitly unset the color so that it doesn't show up.  Since we need a `previewImage` for @jbeezley's suggested use-case, I think it is more consistent to have a `previewColor`.

**Comment by aashish24** [comment]
> If we show color while waiting for tiles to load, then for our standard tile use case, we'll need to explicitly unset the color so that it doesn't show up.

@manthey I don't quite get why we have to unset the color? There is a default color anyways.

**Comment by manthey** [comment]
Right -- we don't *want* a color when tiles don't exist.  For instance, in one of the medical pyramidal tiff images, there are complete tiles for level 3, but on areas that are blank, there are "missing" tiles on level 4 (because they are perfectly represented by the lower-res level 3 tile).  If we draw a default color, this is undesirable because it will hide those layer 3 tiles which are part of the valid overall image.  We could instead set the color to rgba(0,0,0,0), but then we have to check that we don't need to draw pixels instead of just skipping the process, which is computationally wasteful.

**Comment by aashish24** [comment]
> Right -- we don't want a color when tiles don't exist

I see, I am getting what you are saying.  but for previewColor you would have to draw something right? I don't know then if we need previewColor and previewImage. Do we have a use-case?

**Comment by manthey** [comment]
For maps where we *aren't* showing lower level tiles, showing a neutral 'land' color while loading could be desirable.  Or, as @jbeezley said, we can show a grid-like texture.  When we do show lower level tiles, it isn't very useful (unless the tile server is very slow).  For mobile use, showing only the current level and a single preview image for unloaded tiles might be the best behavior to minimize bandwidth and make an informative display.

**Comment by aashish24** [comment]
> For maps where we aren't showing lower level tiles, showing a neutral 'land' color while loading could be desirable. Or, as @jbeezley said, we can show a grid-like texture.

Sounds good to me +1, I didn't see @jbeezley comment, thanks for pointing it out. May be previewImage (or fallback image) could be a mathematical pattern (a function) as well but I don't know if that will complicate things. Although does it mean that by default we will have a grid like image or the default would be just color?

**Comment by manthey** [comment]
By default, I wouldn't have a `previewColor` or `previewImage`.  These would need to be specified for when someone wants them.  Generating an image via a function is far beyond the scope of this PR -- we'll accept a url or Image object.  If someone wants to write a function that generates an Image object, we can obviously take that.

**Comment by aashish24** [comment]
> By default, I wouldn't have a previewColor or previewImage. These would need to be specified for when someone wants them. Generating an image via a function is far beyond the scope of this PR -- we'll accept a url or Image object. If someone wants to write a function that generates an Image object, we can obviously take that.

Sounds good to me :+1: , I was thinking the same. Also, previewColor or previewImage would be nice but not the must have requirement right now. I think the main reason for this work is  performance and flexibility in IMO.

**Comment by manthey** [comment]
When you have a textured quadrilateral, we divide that quad into two triangles, and each triangle is interpolated separately.  This means that if your quadrilateral is not uniform, the two parts can appear disparate.  We split the quadrilateral, arbitrarily, along one diagonal.  You would get a different look along the other diagonal.  For example, the grid pattern below is a distorted square:

![image](https://cloud.githubusercontent.com/assets/8781639/12659719/3d381d3c-c5de-11e5-8647-595b7fcad824.png)

There is no single way to interpolate a quadrilateral which is not symmetric around a diagonal.  I don't think we need to worry about this, but wanted to make sure.

**Comment by aashish24** [comment]
> I don't think we need to worry about this, but wanted to make sure

Roger that. I won't worry about this. I think folks wanted to do different or better will 
have to come up with their own CG algorithm or technique.

**Comment by aashish24** [comment]
By the way, thats a crazy looking quad :smile:

**Comment by manthey** [comment]
For those eager for a preview of this, see the `quad-feature` branch (you'll need a change from vgl, too).  It still needs tests, and the example included there isn't something I want as a primary example (but does show all of the features).

We still allocate an deallocate more memory than I'd like, but I'm getting very good performance with the change.

**Comment by aashish24** [comment]
> We still allocate an deallocate more memory than I'd like, but I'm getting very good performance with the change.

:+1: :+1: how much better?

**Comment by manthey** [comment]
I don't have hard numbers yet, but:

As a test I was loading the transitions examples and running each in turn.

If they have been run once, so that no tiles need to be fetched, before this change I would see three or four dropped frame in the pan and bounce transitions, a half-dozen to a dozen in the elastic and fly transitions, and "lots" of dropped frames in the spin transition.

Now, I get *no* dropped frames from that same test.

**Comment by jbeezley** [comment]
I can report that it is remarkably better.

**Comment by aashish24** [comment]
I am wondering if we can find a way to performance test our code. In fact we need it for many other projects as well.

**Comment by jbeezley** [comment]
I was thinking a bit about how we can go about testing performance.  I want to avoid having tests (at least in our CI workflow) that is hardware dependent or non-deterministic.  A possible test against performance regressions would be to assert an upper bound on the number of objects that are allocated in various synchronous function calls.  Beyond that, there is the idea of collecting the data while testing (#334) so that we can see trendlines and have some data to go by in the review process.

**Comment by manthey** [comment]
Currently, in the quad-feature branch I have an ugly example showing off things you can do with quads.  This is a good example in that it shows off behaviors of the quad feature.  But, I think it is ugly.

Should I leave it in?  We can leave it in source and just not show it on the examples page.

If you think it would improve things, I could put nicer images in it, but otherwise leave it the same (in which case, suggest what the images should be).

Or, I could remove it, in which case we should have a different example of some sort (as the quads are a lot more flexible than the basic tile use shows).

Lastly, we probably should have a selenium test of the quads; just using this example would be sufficient, I think.

I'd appreciate others' opinions and feedback.

**Comment by aashish24** [comment]
> Lastly, we probably should have a selenium test of the quads; just using this example would be sufficient, I think.

+1, yes, I know what we are not encouraging but atleast one test would be nice. 

> f you think it would improve things, I could put nicer images in it, but otherwise leave it the same (in which case, suggest what the images should be).

I think we should remove the bad quads since for new folks, it could be confusing and they might know the context. I will just document in the code and in the example that a wrong order will result in bad looking / wrong quad which should be expected anyways.

**Comment by manthey** [comment]
It is already in the docs (both in the code and such that it appears in the apidocs).

**Comment by aashish24** [comment]
> It is already in the docs (both in the code and such that it appears in the apidocs).

Cool, I would suggest remove the bad quad code then

**Comment by manthey** [comment]
Resolved by adding the 'quad' feature in PR #528.
